modules/transactionpool/update.go

```diff
@@ -387,3 +388,245 @@ func (tp *TransactionPool) PurgeTransactionPool() {
+	// or put in the stack to be removed.
+	removedObjects := make(map[ObjectID]struct{})
+	// removalStack is a stack of objects from removed transactions that
+	// need to be removed from other transactions.
+	var removalStack []ObjectID
+
+	// addTxnToRemovalStack adds all outputs of a transaction into the removal
+	// stack, and marks all fields (including outputs) for removal. This is to
+	// ensure that references to those objects don't remain in the
+	// transactionpool state and incur a memory-leak.
+	addTxnToRemovalStack := func(txn types.Transaction) {
+		// Add all output fields to the set of objects to remove and add
+		// them to the removal stack.
+		for i := range txn.SiacoinOutputs {
+			id := ObjectID(txn.SiacoinOutputID(uint64(i)))
+			if _, alreadyQueuedForRemoval := removedObjects[id]; !alreadyQueuedForRemoval {
+				removedObjects[id] = struct{}{}
+				removalStack = append(removalStack, id)
+			}
+		}
+		for i := range txn.SiafundOutputs {
+			id := ObjectID(txn.SiafundOutputID(uint64(i)))
+			if _, alreadyQueuedForRemoval := removedObjects[id]; !alreadyQueuedForRemoval {
+				removedObjects[id] = struct{}{}
+				removalStack = append(removalStack, id)
+			}
+		}
+
+		// TODO: Make sure all outputs are accounted for properly in file
+		// contract related fields. Do `NewValidProofOutputs` and
+		// `NewMissedProofOutputs` fields of a FileContractRevision need to be
+		// added to the removal stack as well?
+		for i := range txn.FileContracts {
+			id := ObjectID(txn.FileContractID(uint64(i)))
+			if _, alreadyQueuedForRemoval := removedObjects[id]; !alreadyQueuedForRemoval {
+				removedObjects[id] = struct{}{}
+				removalStack = append(removalStack, id)
+			}
+		}
+		for _, fcr := range txn.FileContractRevisions {
+			id := ObjectID(crypto.HashObject(fcr))
+			if _, alreadyQueuedForRemoval := removedObjects[id]; !alreadyQueuedForRemoval {
+				removedObjects[id] = struct{}{}
+				removalStack = append(removalStack, id)
+			}
+		}
+		for _, sp := range txn.StorageProofs {
+			id := ObjectID(crypto.HashObject(sp))
+			if _, alreadyQueuedForRemoval := removedObjects[id]; !alreadyQueuedForRemoval {
+				removedObjects[id] = struct{}{}
+				removalStack = append(removalStack, id)
+			}
+		}
+	}
+
+	// Add the given transaction's objects to the stack, and then handle all
+	// objects in the stack.
+	addTxnToRemovalStack(txn)
+	for n := len(removalStack); n != 0; {
+		// Pop off next object ID from the stack.
+		nextObjectID := removalStack[n-1]
+		removalStack = removalStack[:n-1]
+
+		// Get the location of the object just popped from the stack.
+		// TODO: this needs to be a func that gets locations from inputs that refer to
+		// outputs in the stack
+		nextObjectLocations, ok := objectLocations[nextObjectID]
+		if !ok {
+			nextObjectLocations, ok = tp.findObjectLocations(nextObjectID)
+			if !ok {
+				//TODO: Decide: Is this correct behavior? should this ever happen?
```

Comment:
If an object was marked for removal, and it cannot be found in the tpool then something crazy has probably happened. Not sure what the correct behavior here should be.